File: python/segment.py
n = 100
seg = [0]*(2*n)
def update(i,x):
    seg[n+i] = x
    i = (n+i)//2
    while i:
        seg[i] = max(seg[2*i],seg[2*i+1])
        i >>= 1
def rmq(l,r):
    l += n; r += n
    m = 0
    while l < r:
        #if l&1: m = max(m,seg[l])
        #if r&1: m = max(m,seg[r-1])
        m = max(m,seg[l],seg[r-1])
        l = (l+1)//2; r = r//2
    return m

"""
update(67,3)
update(71,2)
update(45,6)
print( rmq(66,68) )
print( rmq(66,99) )
print( rmq(46,66) )
print( rmq(20,90) )
print( rmq(67,71) )
print( rmq(68,71) )
print( rmq(68,72) )
"""

File: cpp/segment.cpp
using T = int;
const T def = INT_MAX;
T f(T a, T b){ return min(a,b); }

int N;
const int MAX = 200000;
T seg[2*MAX]; //set to default
void update(int i, T v){
    seg[i += N] = v;
    while(i /= 2) seg[i] = f(seg[2*i],seg[2*i+1]);
}
T query(int l, int r){
    l += N, r += N;
    T m1 = def, m2 = def;
    while(l<r){
        if(l&1) m1 = f(m1,seg[l++]);
        if(r&1) m2 = f(seg[--r],m2);
        l /= 2, r /= 2;
    }
    return f(m1,m2);
}

"""
int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n,q; cin >> n >> q;
    N = n;
    for( int i = 0; i < n; i++ ) cin >> seg[i+n];
    for( int i = n-1; i > 0; i-- ) seg[i] = f(seg[2*i],seg[2*i+1]);
    while(q--){
        int o; cin >> o;
        switch(o){
            case 2: int l,r; cin >> l >> r; cout << query(l-1,r) << '\n'; break;
            case 1: int i,x; cin >> i >> x; update(i-1,x); break;
        }
    }
}
"""